process/asyncExecution/headersExecutor.go

```diff
@@ -0,0 +1,98 @@
+package asyncExecution
+
+import (
+	"context"
+
+	"github.com/multiversx/mx-chain-core-go/core/check"
+	"github.com/multiversx/mx-chain-go/process/asyncExecution/queue"
+	logger "github.com/multiversx/mx-chain-logger-go"
+)
+
+var log = logger.GetOrCreate("process/asyncExecution")
+
+// ArgsHeadersExecutor holds all the components needed to create a new instance of *headersExecutor
+type ArgsHeadersExecutor struct {
+	BlocksQueue      BlocksQueue
+	ExecutionTracker ExecutionResultsHandler
+	BlockProcessor   BlockProcessor
+}
+
+type headersExecutor struct {
+	blocksQueue      BlocksQueue
+	executionTracker ExecutionResultsHandler
+	blockProcessor   BlockProcessor
+	cancelFunc       context.CancelFunc
+}
+
+// NewHeadersExecutor will create a new instance of *headersExecutor
+func NewHeadersExecutor(args ArgsHeadersExecutor) (*headersExecutor, error) {
+	if check.IfNil(args.BlocksQueue) {
+		return nil, ErrNilHeadersQueue
+	}
+	if check.IfNil(args.ExecutionTracker) {
+		return nil, ErrNilExecutionTracker
+	}
+	if check.IfNil(args.BlockProcessor) {
+		return nil, ErrNilBlockProcessor
+	}
+
+	return &headersExecutor{
+		blocksQueue:      args.BlocksQueue,
+		executionTracker: args.ExecutionTracker,
+		blockProcessor:   args.BlockProcessor,
+	}, nil
+}
+
+// StartExecution starts a goroutine to continuously process blocks from the queue
+// and add their results to the execution tracker until cancelled or closed.
+func (he *headersExecutor) StartExecution() {
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	he.cancelFunc = cancelFunc
+
+	go he.start(ctx)
+}
+
+func (he *headersExecutor) start(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			// blocking operation
+			headerBodyPair, ok := he.blocksQueue.Pop()
+			if !ok {
+				// close event
+				return
+			}
+
+			he.process(headerBodyPair)
```

Comment:
we might be able to decide at this point if we need to retry or move to process another block (replacement block), if the processing has failed - returns error.